xwiki-platform-core/xwiki-platform-display/xwiki-platform-display-html/src/main/java/org/xwiki/rendering/display/html/internal/DefaultTemplateHTMLDisplayer.java

```diff
@@ -0,0 +1,124 @@
+ * Software Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA
+ * 02110-1301 USA, or see the FSF site: http://www.fsf.org.
+ */
+package org.xwiki.rendering.display.html.internal;
+
+import java.util.HashMap;
+import java.util.Map;
+
+import javax.inject.Inject;
+import javax.inject.Singleton;
+import javax.script.ScriptContext;
+
+import org.xwiki.component.annotation.Component;
+import org.xwiki.displayer.HTMLDisplayer;
+import org.xwiki.displayer.HTMLDisplayerException;
+import org.xwiki.script.ScriptContextManager;
+import org.xwiki.template.TemplateManager;
+
+/**
+ * Default implementation of {@code HTMLDisplayer} using templates.
+ *
+ * @version $Id$
+ * @since 10.11RC1
+ */
+@Component
+@Singleton
+public class DefaultTemplateHTMLDisplayer implements HTMLDisplayer
+{
+    /**
+     * Folder containing the HTML Displayers velocity templates.
+     */
+    public static final String TEMPLATE_FOLDER = "html_displayer";
+
+    /**
+     * Template extension (velocity).
+     */
+    public static final String TEMPLATE_EXTENSION = ".vm";
+
+    @Inject
+    protected TemplateManager templateManager;
+
+    @Inject
+    protected ScriptContextManager scriptContextManager;
+
+    /**
+     * {@inheritDoc}
+     *
+     * Displays the value with the 'view' mode.
+     */
+    @Override
+    public String display(Object value) throws HTMLDisplayerException
+    {
+        return display(value, new HashMap<>());
+    }
+
+    /**
+     * {@inheritDoc}
+     *
+     * Displays the value with the 'view' mode.
+     */
+    @Override
+    public String display(Object value, Map parameters) throws HTMLDisplayerException
+    {
+        return display(value, parameters, "view");
+    }
+
+    @Override
+    public String display(Object value, Map parameters, String mode) throws HTMLDisplayerException
+    {
+        try {
+            ScriptContext scriptContext = scriptContextManager.getCurrentScriptContext();
+            scriptContext.setAttribute("value", value, ScriptContext.ENGINE_SCOPE);
+            scriptContext.setAttribute("parameters", parameters, ScriptContext.ENGINE_SCOPE);
+            scriptContext.setAttribute("mode", mode, ScriptContext.ENGINE_SCOPE);
+
+            return templateManager.render(getTemplateName(value, mode));
+        } catch (Exception e) {
+            throw new HTMLDisplayerException("Couldn't render the template", e);
+        }
+    }
+
+    /**
+     * Computes the template name.
+     * The following names will be use in this priority order to find an existing template:
+     * <ul>
+     *   <li>html_displayer/[type]/[mode].vm
+     *   <li>html_displayer/[type].vm
+     *   <li>html_displayer/[mode].vm
+     *   <li>html_displayer/default.vm
+     * </ul>
+     * @return the template name used to make the rendering
+     */
+    protected String getTemplateName(Object value, String mode) {
+        String type = value.getClass().getSimpleName().toLowerCase();
```

Comment:
This won't work well if the value is null which is not really something exotic.

Reply:
Right, I'll write a check.

Reply:
You might want to separate `AbstractTemplateHTMLDisplayer` which can use the Type indicated by its extending component from `DefaultHTMLDisplayer` which extend `AbstractTemplateHTMLDisplayer` with no explicit Type.

Reply:
@tmortagne I'm not sure to understand. Could you give me an example?

Reply:
I mean that instead of using default HTMLDisplayer as a component other HTML displayers should extend AbstractTemplateHTMLDisplayer so that it knows what is the Type even when the passed value null.

The alternative is to change all the signatures in HTMLDisplayer to also take the Type (and in which case probably get rid of them in HTMLDisplayerManager). I guess another benefit is that it makes multi Type displayers life easier.